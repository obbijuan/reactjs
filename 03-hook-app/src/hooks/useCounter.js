// El objectivo de un custom hooks es la reutilizacion.
import { useState, useEffect } from 'react'

export const useCounter = ( ArrMovies=[], state ) => {
    
    const [listMovies] = useState(ArrMovies);
    const [position, setPosition] = useState(state);
    const [idMovie, setIdMovie] = useState('');
    
    useEffect(() => {
        setIdMovie(listMovies[position]);
        // eslint-disable-next-line
    }, [position])

    function increment() {
        setPosition(position + 1);
    }
    
    function reset(){
        setPosition(0);
    }
    
    const next = () => {
        position < (listMovies.length - 1) ? increment() : reset()
        setIdMovie(listMovies[position]);
    }

    return {
        idMovie,
        position,
        increment,
        next
    }

}
