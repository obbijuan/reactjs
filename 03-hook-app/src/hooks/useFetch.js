import { useState, useEffect } from 'react'

export const useFetch = ( url ) => {
    
    const [state, setState] = useState({ data: null, loading: true })
    
    
    useEffect(() => {
        
        setState({ data: null, loading: true })
        
        fetch(url)
            .then(resp => resp.json())
            .then(data => {
                setState({
                    loading: false,
                    data
                })
            })

    }, [url])

    return state;
    
}
