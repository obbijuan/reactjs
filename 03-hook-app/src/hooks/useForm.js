import { useState } from "react"

// Este custom hook se encarga de manejar los formularios
export const useForm = (initialState = {}) => {
    
    const [values, setValues] = useState(initialState);

    const reset = () => {
        setValues(initialState);
    }

    const handleInputChange = ({ target }) => {
        setValues({
            ...values,
            [target.name] : target.value
        })
    }
    // En el primer valor del arregle devuelve el estado del formulario
    // y el segundo valor el handleInputChange.
    return [values, handleInputChange, reset]
}
