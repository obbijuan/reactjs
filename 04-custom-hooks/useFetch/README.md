# useFetch

Example:
```
    const url = 'endpoint';
    const { data: null, loading: true, error: null } = useFetch(url);

```