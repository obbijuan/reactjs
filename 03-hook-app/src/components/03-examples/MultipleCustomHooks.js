import React, { useEffect, useRef, useState } from 'react'
import { useCounter } from '../../hooks/useCounter'
import '../02-useEffect/effects.css'

export const MultipleCustomHooks = () => {

    const listMovies = ['tt0133093', 'tt0234215', 'tt0242653', 'tt0120737', 'tt0167261',
                        'tt0167260', 'tt0120915', 'tt0121765', 'tt0121766', 'tt8111088'];
    const isMounted = useRef(true);
    const [datos, setDatos] = useState({});
    const { idMovie, next } = useCounter(listMovies);

    useEffect(() => {

        return () => {
            isMounted.current = false;
        }
    }, [])
    
    
    useEffect(() => {
        
        async function fetchData() {
            await fetch(`http://www.omdbapi.com/?i=${idMovie}&apikey=9064e370`)
            .then(resp => resp.json())
                .then(data => {
                    if (isMounted.current) {
                        const { Title, Year, Poster, Plot, Genre } = !!data && data
                        setDatos({ loading: false, Title, Year, Poster, Plot, Genre })
                    }
            })
        }
        fetchData();
    }, [idMovie])
    
    
    return (
        <div>
            <h1>OMDb API</h1>
            <hr />
            <button className="btn btn-primary mb-3" onClick={() => { next();}}>Siguiente</button>
            {
                datos.loading
                ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                :
                    (
                        <blockquote className="blockquote text-left">
                            <p className="mb-0">{datos.Title} - ({datos.Year})</p>
                            <p className="mb-3 h6">{datos.Genre}</p>
                            <img className="mb-3" src={datos.Poster} alt={datos.Title} width="250px"></img>
                            <footer className="blockquote-footer">{datos.Plot}</footer>
                        </blockquote>
                    )
            }
        </div>
    )
}
