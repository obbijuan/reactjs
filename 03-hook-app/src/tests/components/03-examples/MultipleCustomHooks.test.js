import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');

describe('Pruebas en MultipleCustomHooks', () => {
    
    test('debe mostrarse correctamente', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    })

    test('debe mostrar la informacion', () => {

        useFetch.mockReturnValue({
            data: {
                Title: "The Matrix",
                Year: "1999"
            },
            loading: false,
            error: null
        })

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper.find('.alert').exists()).toBe(false);

    })
    

})
