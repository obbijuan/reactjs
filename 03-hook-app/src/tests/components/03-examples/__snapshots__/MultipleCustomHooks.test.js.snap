// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks debe mostrarse correctamente 1`] = `
<div>
  <h1>
    OMDb API
  </h1>
  <hr />
  <button
    className="btn btn-primary mb-3"
    onClick={[Function]}
  >
    Siguiente
  </button>
  <blockquote
    className="blockquote text-left"
  >
    <p
      className="mb-0"
    >
       - (
      )
    </p>
    <p
      className="mb-3 h6"
    />
    <img
      className="mb-3"
      width="250px"
    />
    <footer
      className="blockquote-footer"
    />
  </blockquote>
</div>
`;
